<a href="https://www.kaggle.com/rahultheogre/eda-analyzing-sales-data-to-understand-sale?scriptVersionId=85475112" target="_blank"><img align="left" alt="Kaggle" title="Open in Kaggle" src="https://kaggle.com/static/images/open-in-kaggle.svg"></a>

### Dataset: Online_sales.csv
#### Domain: Retail
##### Objective: Analyzing sales data to understand sales trend

##### 1. Import necessary libraries and read the provided dataset (online_sales.csv) and check the top 5 and random 5 samples of the dataframe. (0.5 point)

1. The "online_sales.csv" file is in the same folder the one in which Python_ProjectNo1 notebook file is.
2. head() method of Pandas DataFrame gives first five rows
3. sample() gives any 5 randome rows.

In [1]:
import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

/kaggle/input/iiitd-project1/online_sales.csv


In [2]:

dataset = pd.read_csv("../input/iiitd-project1/online_sales.csv")

dataset.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

    InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/10 8:26       2.55     17850.0  United Kingdom  
1  12/1/10 8:26       3.39     17850.0  United Kingdom  
2  12/1/10 8:26       2.75     17850.0  United Kingdom  
3  12/1/10 8:26       3.39     17850.0  United Kingdom  
4  12/1/10 8:26       3.39     17850.0  United Kingdom

In [3]:
dataset.sample(5)

InvoiceNo StockCode                          Description  Quantity  \
223588    556484     22326  ROUND SNACK BOXES SET OF4 WOODLAND          2   
231365    557263     23207            LUNCH BAG ALPHABET DESIGN       100   
86280     543543     22722    SET OF 6 SPICE TINS PANTRY DESIGN         1   
33645     539281     22318        FIVE HEART HANGING DECORATION         6   
122500    546877     84946      ANTIQUE SILVER TEA GLASS ETCHED        24   

           InvoiceDate  UnitPrice  CustomerID         Country  
223588   6/12/11 13:17       2.95     16938.0  United Kingdom  
231365   6/19/11 11:08       1.45     14646.0     Netherlands  
86280     2/9/11 15:33       3.95     17843.0  United Kingdom  
33645   12/16/10 15:45       2.95     18176.0  United Kingdom  
122500   3/17/11 16:58       1.25     18210.0  United Kingdom

#### 2. Check info of the dataframe and write your observations. Comment on data types and shape of the dataset. (0.5 point)

In [4]:
#info method gives a summary statistics of the dataframe
dataset.info()
dataset.shape

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 240007 entries, 0 to 240006
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    240007 non-null  object 
 1   StockCode    240007 non-null  object 
 2   Description  239106 non-null  object 
 3   Quantity     240007 non-null  int64  
 4   InvoiceDate  240007 non-null  object 
 5   UnitPrice    240007 non-null  float64
 6   CustomerID   172782 non-null  float64
 7   Country      240007 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 14.6+ MB


(240007, 8)

###### Comments
1. It is a dataset of online sales of a firm. 
2. There are 8 columns, and hence 8 variables through which the data is viewed.
3. There are 240007 rows. The shape is (240007, 8). It is a fairly large dataset.
4. InvoiceNo is an object (str or mixed), which could have come as a surprise, because we would expect it to be numeric. But one, we don't want any kind of accidental change in the InvoiceNo. And two, even though all the sample values seem numeric in the current dataset, we can have alphanumeric/numeric characters in an InvoiceNo.
5. CustomerID is a float64 which seems weird. It could easily have been int64 like Quantity. It occupies more memory space than required.
6. InvoiceDate is an object, which was expected. We often deal with str inputs in columns of datetimes. We would have to convert it into datetime when time comes.  
7. There are many null values in the dataset, specially the CustomerIDs. We may have to drop them. Dropping them won't affect our analysis much, because we have a lot of rows.

#### 3. Check for null values and report the percentage of null values of each column.And drop the rows having null values in it. (1 point)


- STEPS:
1. Use isna() method to check for null values. It returns a mask of bool values for each element in dataset that indicates whether an element is an NA value or not. Each NaN is booled as True i.e. 1.
2. Apply sum() method to the above mask to return a Series of sums of 1's for each column.
3. Convert the Series of [columns, sums of NaNs] into a DataFrame using pd.DataFrame() for better visualisation and manipulation. Call this new dataframe as d_null.
4. Use columns attribute to name the one column d_null has as NoOfNull.
5. Use apply method to add a new column(axis = 1) called PercentNull. 
6. In apply(), use lambda function to perform the required operation on each instance of NoOfNull 
7. Print d_null. That would report the percentage of null values of each column of the dataset

In [5]:
d_null = dataset.isna().sum() #Steps 1,2

d_null = pd.DataFrame(d_null) #Step 3

d_null.columns = ['NoOfNull'] #Step 4

d_null['PercentNull'] = d_null.apply(lambda row: (row.NoOfNull/len(dataset))*100 , axis = 1) #step 5,6

print(d_null) #Step 7

             NoOfNull  PercentNull
InvoiceNo           0     0.000000
StockCode           0     0.000000
Description       901     0.375406
Quantity            0     0.000000
InvoiceDate         0     0.000000
UnitPrice           0     0.000000
CustomerID      67225    28.009600
Country             0     0.000000


###### To drop the rows having null values in it
STEPS:
1. use dropna() method to drop all NaN values.
2. use info() method to display the change
- NOTE: I am making a new DataFrame for this question. In Practice, I would have altered the original dataset with reassignment. But I want the examiners to be able to come back and check, and any alteration in dataset is permanent. 

In [6]:
dataset_NoNull = dataset.dropna() #Step 1
dataset_NoNull.info() #Step 2

<class 'pandas.core.frame.DataFrame'>
Int64Index: 172782 entries, 0 to 240006
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    172782 non-null  object 
 1   StockCode    172782 non-null  object 
 2   Description  172782 non-null  object 
 3   Quantity     172782 non-null  int64  
 4   InvoiceDate  172782 non-null  object 
 5   UnitPrice    172782 non-null  float64
 6   CustomerID   172782 non-null  float64
 7   Country      172782 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 11.9+ MB


##### 5. Drop the instances having quantity less than zero. (1 point) (Note:- refer to the ‘Quantity’ feature)

###### NOTE: I am solving Q 5 before Q 4, because we don't want negative Quantity values in a dataset when performing a statistical analysis on it.

- STEPS
1. Get the indices with negative Quantity instances with by
    - selecting the rows where Quantity<0, and
    - selecting the indices of those rows with index attribute
    - Store the indices in i
2. Use drop() method to remove the selected indices contained in i. Have inplace parameter as False so the original dataset does not change.
3. clean_data is the resultant dataset which has no null values and no rows with negative Quantity values. We will use this dataset from now on.

In [7]:
i = dataset_NoNull[dataset_NoNull['Quantity'] < 0].index #Step 1

clean_data = dataset_NoNull.drop(i, axis = 'rows', inplace = False) #Step 2

clean_data.info() #3 

<class 'pandas.core.frame.DataFrame'>
Int64Index: 168675 entries, 0 to 240006
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    168675 non-null  object 
 1   StockCode    168675 non-null  object 
 2   Description  168675 non-null  object 
 3   Quantity     168675 non-null  int64  
 4   InvoiceDate  168675 non-null  object 
 5   UnitPrice    168675 non-null  float64
 6   CustomerID   168675 non-null  float64
 7   Country      168675 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 11.6+ MB


#### 4. Check the statistical summary of the dataset. (Use inbuilt functions) (0.5 point)

1. I will work on clean_data, which is what we should ideally work with. That is why I decided to solve Q 5 first.
2. REASON (IMPORTANT): 
- About 28% of CustomerId values are NaN. We have to take them as bogus as every Invoice should have a CustomerId. How can a record have an InvoiceNo but no CustomerId? 
- Also, Quantity vaues cannot be negative. So, bogus data. Description does not matter, as it seems  optional. But a very small percentage of data loss would be there (0.37% bogus data). And with so many entries at hand, we can handle that data loss. 

In [8]:
dataset_NoNull.describe() 
#Here min value of Quantity is -74215, which is inadmissible

Quantity      UnitPrice     CustomerID
count  172782.000000  172782.000000  172782.000000
mean       12.244835       3.793566   15274.819941
std       257.898235     101.069930    1725.093177
min    -74215.000000       0.000000   12346.000000
25%         2.000000       1.250000   13842.000000
50%         5.000000       1.950000   15132.000000
75%        12.000000       3.750000   16814.000000
max     74215.000000   38970.000000   18287.000000

#### Solution to Q 4

- STEPS
1. Use descibe() method to give basic descriptive statistical summary. It chooses only those columns which have numeric values as instances.
2. Note: describe() on clean_data gives us more valid values. Case in point: min(Quantity) is now 1

In [9]:
clean_data.describe() #Step 1

Quantity      UnitPrice     CustomerID
count  168675.000000  168675.000000  168675.000000
mean       13.302001       3.285819   15282.751736
std       186.458144      24.203825    1724.910408
min         1.000000       0.000000   12346.000000
25%         2.000000       1.250000   13859.000000
50%         6.000000       1.950000   15147.000000
75%        12.000000       3.750000   16818.000000
max     74215.000000    8142.750000   18287.000000

##### 6. Check unique values of the country and report the name of the country that has the highest number of instances/rows. (0.5 point)

###### Unique values of the country
- STEPS:
1. Extract the Country column using dot operator from the clean_data as a Series
2. Apply unique() method to it to have only unique values in an array unique_values
3. Print to give out Unique values of the country. Use * unpacking operator to get comma separated values.

In [10]:
unique_values = clean_data.Country.unique() #Step 1,2

print("Unique values of the country in the dataset are:\n")
print(*unique_values, sep = ', ') #Step 3

Unique values of the country in the dataset are:

United Kingdom, France, Australia, Netherlands, Germany, Norway, EIRE, Switzerland, Spain, Poland, Portugal, Italy, Belgium, Lithuania, Japan, Iceland, Channel Islands, Denmark, Cyprus, Sweden, Finland, Austria, Greece, Singapore, Lebanon, United Arab Emirates, Israel, Saudi Arabia, Czech Republic, Canada, Unspecified, Brazil, USA, European Community, Bahrain, Malta, Unit


###### Report the name of the country with highest number of instances.
- STEPS
1. Group the dataset according to Country column using groupby method 
2. Aggregate by count() method choosing any column (all will give the same). I choose Description. It will give us a Pandas Series.
3. Convert the resultant Series into a dictionary dict1 using to_dict() Series method
4. Use lambda in max() function to iterate through each value of dict, and get the key (country) with maximum value. Print the result.

In [11]:
dict1 = clean_data.groupby('Country').count().Description.to_dict() #Steps 1,2,3

print(f"The country with highest number of instances is: {max(dict1, key = lambda x: dict1[x])}") #Step 4

The country with highest number of instances is: United Kingdom


##### 7. Create a new column with the name as ‘sales’ having total sales. The total sales is defined as Quantity*UnitPrice. (1 point)b

STEPS:
1. Add a column with Index operator.
2. Assign it values using list comprehension
3. In list comprehension, zip two variables x and y, corresponding to values in columns Quantity and Price and get the result x*y assigned to new column

In [12]:
clean_data['sales'] = [x*y for (x,y) in zip(clean_data['Quantity'],clean_data['UnitPrice'])] #Steps 1,2,3

clean_data.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

    InvoiceDate  UnitPrice  CustomerID         Country  sales  
0  12/1/10 8:26       2.55     17850.0  United Kingdom  15.30  
1  12/1/10 8:26       3.39     17850.0  United Kingdom  20.34  
2  12/1/10 8:26       2.75     17850.0  United Kingdom  22.00  
3  12/1/10 8:26       3.39     17850.0  United Kingdom  20.34  
4  12/1/10 8:26       3.39     17850.0  United Kingdom  20.34

##### 8. Report the top 5 countries in terms of sales. (2 points)

###### a. Considering the size of the sales
- STEPS:
1. Group clean_data by Country using groupby method. Groups attribute is like dictionary whose keys are the computed unique groups and corresponding values being the axis labels belonging to each group. 
2. Aggregate the resulting group with sum()
3. Extract 'sales' column by using dot operator over the aggregate.Now we will have a Series.
4. Sort the Series by value using sort_values() object. Make ascending attribute as False to sort in descending form.
5. Apply index operator to the sorted Series to extract first five index elements, which are country names.

In [13]:
by_sales_size = clean_data.groupby('Country').sum().sales.sort_values(ascending = False).index[0:5] #Steps 1,2,3,4, 5

print("Top five countries in terms of sales size are:")
print(*by_sales_size, sep= ', ')

Top five countries in terms of sales size are:
United Kingdom, Netherlands, Germany, EIRE, France


###### b. Consider the mean value of sales.
- STEPS:
1. Repeat all the steps as in 8(a) except for one change. Aggregate with mean() function instead of sum().

In [14]:
by_sales_mean = clean_data.groupby('Country').mean().sales.sort_values(ascending = False).index[0:5] 

print("Top five countries in terms of mean sales are:")
print(*by_sales_mean, sep= ', ')

Top five countries in terms of mean sales are:
Australia, Netherlands, Japan, Singapore, Sweden


##### 9. Report the top 5 products which bring the highest sales. Use StockCode for product information. (1 point)

- STEPS:
1. Use is_unique attribute on the Series clean_data.StockCode to check if the column has unique values. It turns out to be False. So we have to group clean_data by StockCode first. 
2. Group clean_data by 'StockCode' using groupby method. And aggregate the resulting group with sum().
3. Extract 'sales' column by using dot operator over the aggregate. Now we will have a Series.
4. Sort the Series by value using sort_values() object. Make ascending attribute as False to sort in descending order.
5. Apply index operator to the sorted Series to extract first five index elements, which are Stock codes for top 4 products bringing in highest sales. 

In [15]:
top_5_products = clean_data.groupby('StockCode').sum().sales.sort_values(ascending = False).index[0:5] #Step 2,3,4,5

print("Top five products bringing the highest sales have the StockCodes as:")
print(*top_5_products, sep= ', ')

Top five products bringing the highest sales have the StockCodes as:
22423, 23166, 85123A, 22502, 47566


##### NOTE:
I tried fetching 'Description' values of corresponding 'StockCode' values. They are 'WHITE HANGING HEART T-LIGHT HOLDER', 'PICNIC BASKET WICKER SMALL','REGENCY CAKESTAND 3 TIER', 'PARTY BUNTING' 'MEDIUM CERAMIC TOP STORAGE JAR', 'PICNIC BASKET WICKER 60 PIECES'.

I used the following code:
import numpy as np
top_5_p = clean_data.iloc[np.where(clean_data.StockCode.isin(top_5_products))].Description.unique()
top_5_p

Clearly on of the StockCode values relates to two Description values, most probably 'PICNIC BASKET WICKER SMALL' and 'PICNIC BASKET WICKER 60 PIECES'. So, I left the result to mentioning StockCodes.

##### 10. Convert the ‘InvoiceDate’ into a date format and report the month on which the maximum sales occur? (2 points)

###### Convert the ‘InvoiceDate’ into a date format
- STEPS:
1. Use to_datetime() method over the 'InvoiceDate' column of clean_dataset 

In [16]:
clean_data['InvoiceDate'] = pd.to_datetime(clean_data['InvoiceDate'])
print(f"The object type of column 'InvoiceDate' now convert into date format i.e. {clean_data['InvoiceDate'].dtypes}")

The object type of column 'InvoiceDate' now convert into date format i.e. datetime64[ns]


##### report the month on which the maximum sales occur
STEPS:
1. Create a new column 'month' from 'InvoiceDate' using DatetimeIndex.strftime() method.
    - DatetimeIndex.strftime() gives out an Index of string values of a format supplied to it.
    - Put in %B as parameter to force out full month name. 
2. Group the dataset by 'month' and Aggregate with sum() and save the dataframe in month_max.
3. Extract 'sales' from month_max using dot operator. Now it becomes a Pandas Series.
4. Sort the values in the Series of (months,sales). In descending order because we need the month with maximum sales.
5. Use Index operator on the ordered Series to get the first value- which is what we were after.

In [17]:
clean_data['month'] = pd.DatetimeIndex(clean_data['InvoiceDate']).strftime("%B") #Step 1

month_max = clean_data.groupby('month').sum().sales.sort_values(ascending = False).index[0] #Steps 2,3,4,5

print(f"The month on which maximum sales occur is {month_max}")

The month on which maximum sales occur is May
